core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -342,6 +342,29 @@ class DAGSchedulerSuite
     assert(locs === Seq(Seq("hostA", "hostB"), Seq("hostB", "hostC"), Seq("hostC", "hostD")))
   }
 
+  /**
+   * +---+ shuffle +---+    +---+    +---+
+   * | A |<--------| B |<---| C |<---| D |
+   * +---+         +---+    +---+    +---+
+   * Here, D has one-to-one dependencies on C. C is derived from A by performing a shuffle
+   * and then a map. If we're trying to determine which ancestor stages need to be computed in
+   * order to compute D, we need to figure out whether the shuffle A -> B should be performed.
+   * If the RDD C, which has only one ancestor via a narrow dependency, is cached, then we won't
+   * need to compute A, even if it has some unavailable output partitions. The same goes for B:
+   * if B is 100% cached, then we can avoid the shuffle on A.
```

Comment:
Josh's comment was an awesome description of how the dependencies should be computed, but isn't quite appropriate here as the comment for the test.  What about something like:

This test ensures that if a particular RDD is cached, RDDs earlier in the dependency chain are not computed.  It constructs the following chain of dependencies:
+---+ shuffle +---+    +---+    +---+
 | A |<--------| B |<---| C |<---| D |
+---+         +---+    +---+    +---+
Here, B is derived from A by performing a shuffle, C has a one-to-one dependency on B, and D similarly has a one-to-one dependency on C.  If none of the RDDs were cached, this set of RDDs would result in a two stage job: one ShuffleMapStage, and a ResultStage that reads the shuffled data from RDD A.  This test ensures that if C is cached, the scheduler doesn't perform a shuffle, and instead computes the result using a single ResultStage that reads C's cached data.